Track particle time and implement filter for time

paulromano opened the pull request:
Up to now, particles being transported in OpenMC know nothing about time. This is mostly due to the origin of the code being focused on static k-eigenvalue problems, where time is not really a factor. However, with the increasing usage of the code in other areas, it would be beneficial to have an ability to track the particle's time and filter tallies based on the observed time. This PR extends the `Particle` and `SourceSite` data structures to keep track of time and adds a new `TimeFilter` that allows both volume and surface tallies to filter events based on the particle's clock. The calculation of the particle's velocity is done using relativistic equations so time dilation effects are accounted for.

@cfichtlscherer @ojschumann Let me know what you guys think about the changes here and if they're in line with what you've done previously in your own work.

@py1sl The issue [you mentioned](https://openmc.discourse.group/t/is-particle-lifetime-tracked/1407/9?u=paulromano) with surface tallies should be fixed now. One of the unit tests covers this case (tallying current on a purely absorbing surface a given distance from a monoenergetic, monodirectional point source).

paulromano commented:
One last note -- I had to update some test results that depend on the inverse-velocity score. That score used to rely on non-relativistic formulas, so the switch to using relativistic formulas changed those results ever so slightly.

ojschumann commented on src/tallies/filter.cpp:
```diff
@@ -151,6 +152,8 @@ Filter* Filter::create(const std::string& type, int32_t id)
     return Filter::create<SpatialLegendreFilter>(id);
   } else if (type == "sphericalharmonics") {
     return Filter::create<SphericalHarmonicsFilter>(id);
+  } else if (type == "time") {
```
Perhaps one can refactor the code (later) and make the strings static variables of the classes. 
If there is only one definition of the actual string, no typos can occure.

paulromano replied:
Good point -- we do have a `type()` member function which gives that string, but it requires that you already have an instance. With a little refactoring this could be cleaned up, but let's leave that for another time if it's alright with you.

ojschumann commented on src/tallies/filter_time.cpp:
```diff
@@ -0,0 +1,97 @@
+  bins_.reserve(bins.size());
+
+  // Copy bins, ensuring they are valid
+  for (gsl::index i = 0; i < bins.size(); ++i) {
```
Idea from @gridley to use `std::is_sorted`. From mesh.cpp:

`if (not std::is_sorted(bins.begin(), bins.end(), std::less_equal<double>()))`

paulromano replied:
Addressed in 7b43bb97a5c51a5f8a30f1554fac966442795854

ojschumann commented on src/tallies/filter_time.cpp:
```diff
@@ -0,0 +1,97 @@
+{
+  // Get the start/end time of the particle for this track
+  auto t_start = p.time_last();
+  auto t_end = p.time();
```
`const auto`

and the same for most variables futher down

paulromano replied:
Addressed in 7b43bb97a5c51a5f8a30f1554fac966442795854

ojschumann commented on src/tallies/filter_time.cpp:
```diff
@@ -0,0 +1,97 @@
+
+      // Add match with weight equal to the fraction of the time interval within
+      // the current time bin
+      if (dt_total > 0.0) {
```
compare dt_total to zero (or alternatively FP_PRECISION) and exit prior to the loop.

ojschumann replied:
So; I had another check with my actual source code.

In addition to your changes, I have two writes in Particle::write_restart() 
   write_dataset(file_id, "time", simulation::source_bank[i-1].time);
   write_dataset(file_id, "time", simulation::source_bank[i-1].time);

and a time-distribution for source.cpp and source.py (I can provide a PR, after yours is merged)
Otherwise, my changes are (more or less) identical.

paulromano replied:
Thanks for the careful review! I've added those changes for particle restarts (and the check for dt_total before the loop). I'll leave the source-related changes to you for a future PR

ojschumann reviewed (changes requested):
I will performe some comparisions with MCNP after the holidays

paulromano commented on src/particle.cpp:
```diff
@@ -170,10 +198,11 @@ void Particle::event_advance()
   for (int j = 0; j < n_coord(); ++j) {
     coord(j).r += distance * coord(j).u;
   }
+  this->time() += distance / this->speed();
```
Generally, I prefer to use `this->` to make clear that a member function is being called (as opposed to a non-member function). However, I see that in this function, `this->` is not used for any of the other calls to member functions so this one does stick out -- if you'd prefer, I can remove it.

paulromano commented on src/particle.cpp:
```diff
@@ -36,6 +36,30 @@
+  }
+
+  // Calculate inverse of Lorentz factor
+  double inv_gamma = mass / (this->E() + mass);
```
Addressed in 7b43bb97a5c51a5f8a30f1554fac966442795854

ojschumann reviewed (approved):
I'm not sure if I might approve at all....

paulromano commented:
@ojschumann Yes, your approval is appreciated! I just invited you to the "Committers" team which will give you merge rights (per our [developer's guide](https://docs.openmc.org/en/latest/devguide/contributing.html#becoming-a-committer), anyone who makes non-trivial contributions is given commit rights).

I made one more small change here, which is to remove `less_equal` from the call to `std::is_sorted`. It turns out that duplicates are handled fine for the default comparison operator and `less_equal` can cause problems (see discussion [here](https://github.com/xtensor-stack/xtensor/issues/2296#issuecomment-1001552766)). You'll probably want to make a similar change in #1931.

cfichtlscherer commented on include/openmc/particle_data.h:
```diff
@@ -227,9 +228,11 @@ class ParticleData {
+  double wgt_ {1.0};       //!< particle weight
+  double mu_;              //!< angle of scatter
+  double time_ {0.0};      //!< time in [s]
+  double time_last_ {0.0}; //!< previous time in [s]
```
A more general comment (probably due to my lack of knowledge about certain structures): Why do we take the detour via time_last and time? Can't we just do

time = time + distance / speed.

There is a similar structure when it comes to E and E_last_ which I used for the PHT. Maybe soon we could use it to access the times between interactions, or what is the idea here?

paulromano replied:
For tallies that occur after a particle collision, the speed of the particle has already changed and hence you can't backtrack like this to figure out what time the particle started its track at.

cfichtlscherer replied:
thank you

cfichtlscherer commented on src/particle.cpp:
```diff
@@ -36,6 +36,30 @@
+  }
+
+  // Calculate inverse of Lorentz factor
+  const double inv_gamma = mass / (this->E() + mass);
```
Pure formality, I would define gamma normally here (i.e. not inverse) and write a bit more to make it clear faster.

paulromano replied:
I did it this way so that the same formula works for massless particles (photons). If you do it the "normal" way, it would result in a divide-by-zero for photons.

cfichtlscherer replied:
yes, makes sense

py1sl commented:
Surface tallies look good,
I'm not sure for volume tallies it is right in all cases, but it could be an interpretation problem.
What question is a time filter in a volume seeking to answer? 

- is it of the total result in this volume what fraction is in this time bin - I believe this is the current implementation 
- or is it in time bin x what is the result in this cell?

I think many people would expect the later.
Will try to do some more testing.

ojschumann commented:
@paulromano I checked duplicates with std::is_sorted and the default comparision operator. Unfortunately they are not found. See [here](https://pastebin.com/GjkMJYkV). The best I could come up with is `(std::adjacent_find(D.begin(), D.end(), std::greater_equal<>()) == D.end())`

paulromano commented:
> What question is a time filter in a volume seeking to answer?

@py1sl The time filter I've implemented here seeks to limit the scoring events to a tally to only those that fall within a given time bin. In the case of surface tallies (or a collision estimator for volume tallies), the decision is binary -- either the particle scores, or it doesn't score. However, in the case of a tracklength estimator, the particle may have started in one bin and ended up in a different bin. This is why in that case I calculate the fraction of the current track that falls in a given time bin and limit the scoring  to that fraction of the tracklength rather than the entire tracklength. Applied to, e.g., a tally with a cell filter and the "flux" score, it will still tell you what the flux is in a given cell over a given time bin. If you integrate it over all time, you should just get the total flux.

@ojschumann thanks for following up -- I've just updated the check to use `adjacent_find` as you suggested.

ojschumann commented:
@paulromano Sorry, I made a typo: it has to be `(std::adjacent_find(D.begin(), D.end(), std::greater_equal<>()) != D.end())` as `std::greater_equal<>()` should **never** be true at any position. `std::adjacent_find` thus should not find a position.

In addition, I made some comparisons with MCNP. I will make a post in discourse.
